service/history/workflow/mutable_state_impl.go

```diff
@@ -5113,26 +5113,18 @@ func (ms *MutableStateImpl) closeTransactionHandleActivityUserTimerTasks(
 	}
 }
 
-func (ms *MutableStateImpl) closeTransactionCollapseUpsertVisibilityTasks() {
-	// check if we have >= 2 tasks that are identical upsert visibility tasks
-	// note that VisibilityTimestamp and TaskID are not assigned yet
+func (ms *MutableStateImpl) closeTransactionCollapseVisibilityTasks() {
 	visTasks := ms.InsertTasks[tasks.CategoryVisibility]
 	if len(visTasks) < 2 {
 		return
 	}
-	var task0 *tasks.UpsertExecutionVisibilityTask
-	for i, task := range visTasks {
-		task, ok := task.(*tasks.UpsertExecutionVisibilityTask)
-		if !ok {
-			return
-		}
-		if i == 0 {
-			task0 = task
-		} else if *task != *task0 {
-			return
+	// Visibility tasks are collapsed into a single one: START < UPSERT < COLSE < DELETE
+	// Their enum values are already in order, so using them to make the code simpler.
```

Comment:
is it possible that if someone in the future adds a new visibility task type that can't be collapsed? I think it will be safer if we can check the task type are the ones we know can be collapsed.